indispensable-asn1/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/Asn1Elements.kt

```diff
@@ -474,31 +467,180 @@ sealed class Asn1Structure(
+     *     if (hasNext()) {
+     *         // ...
+     *     }
+     *     DecodedStructure(...)
+     * }
+     * ```
+     *
+     * Manual iteration:
+     * ```kotlin
+     * val iterator = structure.iterator()
+     * if (iterator.hasNext()) {
+     *     // ...
+     * }
+     * ```
      */
-    fun hasMoreChildren() = children.size > index
+    @Deprecated(
+        message = "Use Iterator inside decodeAs/decodeRethrowing",
+        level = DeprecationLevel.ERROR
+    )
+    fun hasMoreChildren() = iterator.hasNext()
 
     /**
      * Returns the current child or `null`, if there are no children left
      * (useful when iterating over this structure's children).
+     * ### Migration Examples:
+     *
+     * Using `decodeRethrowing`:
+     * ```kotlin
+     * structure.decodeRethrowing {
+     *     val decodedChild = peek()?.let {
+     *         // Inspect or conditionally consume the child
+     *         val child = next()
+     *         // ...
+     *     }
+     *     DecodedStructure(...)
+     * }
+     * ```
+     *
+     * Manual iteration:
+     * ```kotlin
+     * val iterator = structure.iterator()
+     * val child = iterator.peek()
+     * if (child != null) {
+     *     // Inspect child without consuming it
+     * }
+     * ```
+     */
+    @Deprecated(
+        message = "Use Iterator inside decodeAs/decodeRethrowing",
+        level = DeprecationLevel.ERROR
+    )
+    fun peek() = iterator.peek()
+
+    /**
+     * An iterator over a list of [Asn1Element] children within an ASN.1 structure
+     * Designed for traversing ASN.1 components in decoding/parsing scenarios
+     * */
+    inner class Iterator(
+        private val children: List<Asn1Element>,
+        private var index: Int = 0
+    ): kotlin.collections.Iterator<Asn1Element> {
+
+        /**
+         * Returns the next child held by this structure. Useful for iterating over its children when parsing complex structures.
+         * @throws [NoSuchElementException] if no more children are available
+         */
+        override fun next() =
+            catching { children[index++] }.getOrElse { throw NoSuchElementException("No more content left") }
+
+        /**
+         * Exception-free version of [next]
+         */
+        fun nextOrNull() = catchingUnwrapped { next() }.getOrNull()
+
+        /**
+         * Returns `true` if more children can be retrieved by [next]. `false` otherwise
+         */
+        override fun hasNext() = children.size > index
+
+        /**
+         * Returns the current child or `null`, if there are no children left
+         * (useful when iterating over this structure's children).
+         */
+        fun peek() = if (!hasNext()) null else children[index]
+
+        /**
+         * Returns iterator with reversed view
+         * */
+        fun reversed(): Iterator =
+            Iterator(children.reversed(), children.lastIndex - index)
+    }
+
+    override operator fun iterator() = Iterator(children)
```

Comment:
why are we passing `children` into `Iterator`'s constructor if `Iterator` is an inner class anyway?